pandas/core/tools/datetimes.py

```diff
@@ -125,11 +125,14 @@ class FulldatetimeDict(YearMonthDayDict, total=False):
 # ---------------------------------------------------------------------
 
 
-def _guess_datetime_format_for_array(arr, dayfirst: bool | None = False):
-    # Try to guess the format based on the first non-NaN element
+def _guess_datetime_format_for_array(arr, dayfirst: bool | None = False) -> str | None:
+    # Try to guess the format based on the first non-NaN element, return None if can't
     non_nan_elements = notna(arr).nonzero()[0]
     if len(non_nan_elements):
-        return guess_datetime_format(arr[non_nan_elements[0]], dayfirst=dayfirst)
+        if type(first_non_nan_element := arr[non_nan_elements[0]]) is str:
```

Comment:
is the fallback path for this less performant?  could we instead cast np.str_ objs to str here?

Reply:
good call, have opened https://github.com/pandas-dev/pandas/issues/48974 for now, will get back to it when I get a chance